compiler/grainformat/reformat.re

```diff
@@ -0,0 +1,2906 @@
+open Grain;
+open Compile;
+open Grain_parsing;
+open Grain_utils;
+
+module Doc = Res_doc;
+
+let exception_primitives = [|"throw", "fail", "assert"|];
+let list_cons = "[...]";
+
+type error =
+  | Illegal_parse(string)
+  | Unsupported_syntax(string);
+
+exception Error(error);
+
+type sugared_list_item =
+  | Regular(Grain_parsing.Parsetree.expression)
+  | Spread(Grain_parsing.Parsetree.expression);
+
+type sugared_pattern_item =
+  | RegularPattern(Grain_parsing.Parsetree.pattern)
+  | SpreadPattern(Grain_parsing.Parsetree.pattern);
+
+let get_raw_pos_info = (pos: Lexing.position) => (
+  pos.pos_fname,
+  pos.pos_lnum,
+  pos.pos_cnum - pos.pos_bol,
+  pos.pos_bol,
+);
+
+let get_original_code =
+    (location: Grain_parsing.Location.t, source: array(string)) => {
+  let (_, startline, startc, _) = get_raw_pos_info(location.loc_start);
+  let (_, endline, endc, _) = get_raw_pos_info(location.loc_end);
+
+  let text = ref("");
+  if (Array.length(source) > endline - 1) {
+    if (startline == endline) {
+      let full_line = source[startline - 1];
+
+      let without_trailing = Str.string_before(full_line, endc);
+
+      text := text^ ++ without_trailing;
+    } else {
+      for (line in startline - 1 to endline - 1) {
+        if (line + 1 == startline) {
+          text := text^ ++ Str.string_after(source[line], startc) ++ "\n";
+        } else if (line + 1 == endline) {
+          text := text^ ++ source[line];
+        } else {
+          text := text^ ++ source[line] ++ "\n";
+        };
+      };
+    };
+    text^;
+  } else {
+    "/* Formatter error*/";
+  };
+};
+
+let last_file_loc = (): Grain_parsing.Location.t => {
+  let endpos: Stdlib__lexing.position = {
+    pos_fname: "",
+    pos_lnum: max_int,
+    pos_cnum: max_int,
```

Comment:
Just trying to get a line number larger than the last line.  Now I have the original source, I can set this precisely!

Reply:
Actually, I can't.  I think this will be a sufficiently large number as we can't process a source file with more lines than this anyway as other things will break.

Reply:
As long as it doesn't wrap in JS and be negative.... @ospencer any insights on this in the JSOO context?